docs/config-merge-logic.md

```diff
@@ -0,0 +1,609 @@
+        "string2": "goodbye"
+    }
+```
+
+Now a client wants to update it by changing `["int1"]` to 1, and adding a new key `["int2"]` set to
+2, and deleting the `"int0"` key.  This new message then gets seqno 123; the full update then looks
+like this:
+
+```json
+    {
+        "#": 123,
+        "&": {  // The new full data
+            "dictB": {
+                "changed": -1,
+                "foo": 123,
+                "removed": "x",
+                "removed2": "y"
+            },
+            "dictC": {
+                "x": {
+                    "y": 1
+                }
+            },
+            "good": [99, 456, "bar"],
+            "great": [-42, "omg"],
+            "int1": 1,
+            "int2": 2,
+            "string1": "hello",
+            "string2": "goodbye"
+        },
+        "*": {
+            "<": [
+                [119, "(hash119)", {...changes-in-seqno-119...}],
+                [120, "(hash120)", {...changes-in-seqno-120...}],
+                [121, "(hash121)", {...changes-in-seqno-121...}],
+                [122, "(hash122)", {...changes-in-seqno-122...}],
+            ],
+            "=": {
+                "int0": "-", // removed
+                "int1": "",  // changed
+                "int2": ""   // added
+            }
+        }
+    }
+```
+
+
+## Large, but still ordinary, update
+
+Suppose an update begins from the following data (with seqno 123), and updates have all been linear
+and orderly (i.e. there have been no recent config conflicts):
+
+```json
+    {
+        "dictB": {
+            "changed": -1,
+            "foo": 123,
+            "removed": "x",
+            "removed2": "y"
+        },
+        "dictC": {
+            "x": {
+                "y": 1
+            }
+        },
+        "good": [99, 456, "bar"],
+        "great": [-42, "omg"],
+        "int1": 1,
+        "int2": 2,
+        "string1": "hello",
+        "string2": "goodbye"
+    }
+```
+
+The client modifies it quite substantially to the following:
+
+```json
+    {
+        "dictA": {
+            "hello": 123,
+            "goodbye": [123, 456]
+        },
+        "dictB": {
+            "added": 9999,
+            "changed": 1,
+            "foo": 123,
+            "nested": {"a", 1}
+        },
+        "good": [99, 123, "Foo", "bar"],
+        "int1": 42,
+        "int2": 2,
+        "string2": "hello",
+        "string3": "omg"
+    }
+```
+
+The overall record of this change looks as follows.
+
+```jason
+    {
+        "#": 124,
+        "&": {  // The current full data
+            "dictA": {
+                "hello": 123,
+                "goodbye": [123, 456]
+            },
+            "dictB": {
+                "added": 9999,
+                "changed": 1,
+                "foo": 123,
+                "nested": {
+                    "a": 1
+                }
+            },
+            "good": [99, 123, "Foo", "bar"],
+            "int1": 42,
+            "int2": 2,
+            "string2": "hello"
```

Comment:
If I understand correctly, we are missing `"string3": "omg"` here?

Reply:
I spotted this and some other errors in here when using these as test cases, which are updated.